File: etc/gh/queue.py
from collections import deque

class Queue:
    def __init__(self):
        self.queue = deque()

    def enqueue(self, item):
        self.queue.append(item)

    def dequeue(self):
        if not self.is_empty():
            return self.queue.popleft()
        return None

    def peek(self):
        if not self.is_empty():
            return self.queue[0]
        return None

    def is_empty(self):
        return len(self.queue) == 0

    def size(self):
        return len(self.queue)

# 사용 예시
if __name__ == "__main__":
    q = Queue()
    q.enqueue(1)
    q.enqueue(2)
    print(q.dequeue())  # 출력: 1
    print(q.peek())  # 출력: 2
    print(q.is_empty())  # 출력: False

File: Data Structure/data_structure/queue.c
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>

#define MAX 100

typedef struct Queue {
    int items[MAX];
    int front, rear;
} Queue;

Queue* create_queue() {
    Queue* queue = (Queue*)malloc(sizeof(Queue));
    queue->front = 0;
    queue->rear = 0;
    return queue;
}

int is_full(Queue* queue) {
    return queue->rear == MAX;
}

int is_empty(Queue* queue) {
    return queue->rear == queue->front;
}

void enqueue(Queue* queue, int item) {
    if (is_full(queue))
        return;
    queue->items[queue->rear++] = item;
}

int dequeue(Queue* queue) {
    if (is_empty(queue))
        return INT_MIN;
    return queue->items[queue->front++];
}

int peek(Queue* queue) {
    if (is_empty(queue))
        return INT_MIN;
    return queue->items[queue->front];
}

int main() {
    Queue* queue = create_queue();
    enqueue(queue, 10);
    enqueue(queue, 20);
    printf("%d dequeued from queue\n", dequeue(queue));  // 출력: 10 dequeued from queue
    printf("Front element is %d\n", peek(queue));        // 출력: Front element is 20
    return 0;
}
